File: src/types/Quotation.ts
export type QuotationStatus = 'Draft' | 'Approved' | 'Rejected' | 'Expired'

export interface QuotationLineItem {
  id: string
  description: string
  quantity: number
  unitPrice: number
}

export interface QuotationRevision {
  id: string
  revision: string
  date: string
  changedBy: string
  summary: string
}

/** One row from the quotation's audit trail (see backend
 * audit_service.get_history) -- document-lifecycle and status-change
 * events like "Quotation emailed"/"Quotation approved", not content
 * edits (those are QuotationRevision, above). Shown merged with
 * revisions in QuotationRevisionHistory.vue. */
export interface QuotationAuditEvent {
  id: string
  action: string
  user: string
  timestamp: string
  previousValue?: string
  newValue?: string
  reason?: string
}

export interface Quotation {
  id: string
  projectId: string
  quotationNo: string
  revision: string
  issueDate: string
  validity: string
  status: QuotationStatus
  currency: string
  preparedBy: string
  discountAmount: number
  notes: string
  termsAndConditions: string[]
  scopePhases: string[]
  paymentTerms: string[]
  lineItems: QuotationLineItem[]
  amount: number
  // undefined while an editable draft; set once finalized.
  finalizedAt?: string
  revisions: QuotationRevision[]
}

File: backend/app/models/quotation.py
from datetime import date, datetime

from sqlalchemy import JSON, Date, DateTime, Enum, ForeignKey, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.mixins import SoftDeleteMixin, TimestampMixin
from app.models.user import BigPK

QUOTATION_STATUSES = ("Draft", "Approved", "Rejected", "Expired")


class Quotation(Base, TimestampMixin, SoftDeleteMixin):
    __tablename__ = "quotations"

    id: Mapped[int] = mapped_column(BigPK, primary_key=True)
    quotation_no: Mapped[str] = mapped_column(String(20), unique=True, nullable=False)
    project_id: Mapped[int] = mapped_column(
        BigPK, ForeignKey("projects.id", ondelete="RESTRICT"), nullable=False, index=True
    )
    revision: Mapped[str] = mapped_column(String(10), nullable=False, default="R0")
    issue_date: Mapped[date] = mapped_column(Date, nullable=False)
    validity: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(
        Enum(*QUOTATION_STATUSES, name="quotation_status"), nullable=False, default="Draft"
    )
    currency: Mapped[str] = mapped_column(String(10), nullable=False, default="KWD")
    prepared_by: Mapped[int] = mapped_column(BigPK, ForeignKey("users.id", ondelete="RESTRICT"), nullable=False)
    discount_amount: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    terms_and_conditions: Mapped[list] = mapped_column(JSON, nullable=False, default=list)
    # Recomputed from line items + discount on every write (see
    # services/quotation_service.py) -- stored rather than computed at
    # read time purely so list views can sort/filter on it without
    # joining and summing line items every time.
    amount: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    # NULL while still an editable draft; set once the user clicks
    # Save as Final, after which content is locked and the document is
    # ready to print. A quotation can't leave Draft status until this
    # is set (see quotation_service.set_status).
    finalized_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)


class QuotationRevision(Base):
    """Mirrors ContractRevision (see models/contract.py) -- one row per
    saved change to a quotation's content, written automatically by
    quotation_service (on create, and on every content-changing update),
    not just via an explicit user action."""

    __tablename__ = "quotation_revisions"

    id: Mapped[int] = mapped_column(BigPK, primary_key=True)
    quotation_id: Mapped[int] = mapped_column(
        BigPK, ForeignKey("quotations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    revision: Mapped[str] = mapped_column(String(10), nullable=False)
    revised_at: Mapped[date] = mapped_column(Date, nullable=False)
    changed_by: Mapped[int] = mapped_column(BigPK, ForeignKey("users.id", ondelete="RESTRICT"), nullable=False)
    summary: Mapped[str] = mapped_column(Text, nullable=False)


class QuotationLineItem(Base):
    __tablename__ = "quotation_line_items"

    id: Mapped[int] = mapped_column(BigPK, primary_key=True)
    quotation_id: Mapped[int] = mapped_column(
        BigPK, ForeignKey("quotations.id", ondelete="CASCADE"), nullable=False, index=True
    )
    description: Mapped[str] = mapped_column(String(300), nullable=False)
    quantity: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    unit_price: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
